adapters/lora-vertx/src/main/java/org/eclipse/hono/adapter/lora/providers/ActilityBaseProvider.java

```diff
@@ -0,0 +1,115 @@
+    protected static final String FIELD_ACTILITY_LRRS = "Lrrs";
+    protected static final String FIELD_ACTILITY_SPREADING_FACTOR = "SpFact";
+    private static final String FIELD_ACTILITY_DEVICE_EUI = "DevEUI";
+    private static final String FIELD_ACTILITY_ROOT_OBJECT = "DevEUI_uplink";
+    private static final String FIELD_ACTILITY_PAYLOAD = "payload_hex";
+
+    protected Optional<JsonObject> getRootObject(final JsonObject loraMessage) {
+        return LoraUtils.getChildObject(loraMessage, FIELD_ACTILITY_ROOT_OBJECT, JsonObject.class);
+    }
+
+    @Override
+    protected String getDevEui(final JsonObject loraMessage) {
+
+        Objects.requireNonNull(loraMessage);
+        return getRootObject(loraMessage)
+                .map(root -> root.getValue(FIELD_ACTILITY_DEVICE_EUI))
+                .filter(String.class::isInstance)
+                .map(String.class::cast)
+                .orElseThrow(() -> new LoraProviderMalformedPayloadException("message does not contain String valued device EUI property"));
+    }
+
+    @Override
+    protected Buffer getPayload(final JsonObject loraMessage) {
+        Objects.requireNonNull(loraMessage);
+        return getRootObject(loraMessage)
+                .map(root -> root.getValue(FIELD_ACTILITY_PAYLOAD))
+                .filter(String.class::isInstance)
+                .map(String.class::cast)
+                .map(s -> Buffer.buffer(BaseEncoding.base16().decode(s.toUpperCase())))
+                .orElseThrow(() -> new LoraProviderMalformedPayloadException("message does not contain String valued payload property"));
+    }
+
+    @Override
+    protected LoraMessageType getMessageType(final JsonObject loraMessage) {
+        Objects.requireNonNull(loraMessage);
+        return getRootObject(loraMessage)
+                .map(root -> LoraMessageType.UPLINK)
+                .orElse(LoraMessageType.UNKNOWN);
+    }
+
+    /**
+     * Gets the frequency corresponding to the channel ID used by Actility/ThingWork
+     * as described in section 2.4 of the
+     * <a href="https://partners.thingpark.com/sites/default/files/2017-11/AdvancedThingParkDeveloperGuide_V4.pdf">
+     * Advanced Developer Guide</a>.
+     *
+     * @param logicalChannelId The channel ID.
+     * @return The frequency in MHz or {@code null} if the identifier is unknown.
+     */
+    protected Double getFrequency(final String logicalChannelId) {
+        switch (logicalChannelId) {
+        case "LC1":
+            return 868.1;
+        case "LC2":
+            return 868.3;
+        case "LC3":
+            return 868.5;
+        case "RX2":
+            return 869.525;
+        default:
+            return null;
+        }
+    }
+
+    @Override
+    protected LoraMetaData getMetaData(final JsonObject loraMessage) {
+
+        Objects.requireNonNull(loraMessage);
+
+        return getRootObject(loraMessage)
+            .map(this::extractMetaData)
+            .orElse(null);
+    }
```

Comment:
Are these methods supposed to be overridden by subclasses? If not then I would suggest to declare them final in order to prevent hard to find problems when somebody does ...

Reply:
no they are not, changed it according to your proposal